### PR TITLE
add fusion_volume functional tests

### PR DESCRIPTION
**Pull request by dimleevik**
##### SUMMARY

Added functional tests for fusion_volume

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

tests

### CODE REVIEW

**Review by sdodsley — DISMISSED** [review]

File: plugins/modules/fusion_volume.py [hunk 1]
```diff
@@ -465,18 +377,11 @@ def validate_arguments(module, fusion):
...
-    if state == "present" and not volume:
-        # would create a volume; check size is lower than storage class limit
-        # (let resize checks be handled by the server since they are more complicated)
-        size = parse_number_with_metric_suffix(module, module.params["size"])
-        size_limit = get_storage_class(module, fusion).size_limit
-        if size > size_limit:
-            module.fail_json(
-                msg="Requested volume size {0} exceeds the storage class limit of {1}".format(
-                    module.params["size"],
-                    print_number_with_metric_suffix(size_limit),
```
[the remark this answers is not in the document]
Comment (dimleevik) [reply plugins/modules/fusion_volume.py:477]: Yes, good catch. Didn't notice that some related functions could be unused at all

File: plugins/modules/fusion_volume.py [hunk 2]
```diff
@@ -465,18 +378,11 @@ def validate_arguments(module, fusion):
...
-                )
-            )
+    size = parse_number_with_metric_suffix(module, module.params["size"])
```
Comment (jankodera-ps) [inline plugins/modules/fusion_volume.py:381]: This gets run even if `module.params["state"] == "absent"`, no? In that case size is optional and parsing and checks probably shouldn't be done

Comment (dimleevik) [reply plugins/modules/fusion_volume.py:381]: But for `destroyed=True` we have order as described on line 316->323, first we update all volume parameters and then only volume become destroyed. So this validation is required even for `state="absent"`

Comment (andreepyro) [reply plugins/modules/fusion_volume.py:381]: Maybe put `if module.params["size"] is not None:` to make sure it doesn't fail if no `size` parameter is provided.

**Review by sdodsley — APPROVED** [review]

File: tests/functional/test_fusion_volume.py [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+def volume(storage_class, placement_group, protection_policy, host_access_policies):
+    return {
+        "name": "volume_1",
+        "display_name": "Volume 1",
+        "tenant": "t1",
+        "tenant_space": "ts1",
+        "storage_class": purefusion.StorageClass(**storage_class),
+        "placement_group": purefusion.PlacementGroup(**placement_group),
+        "protection_policy": purefusion.ProtectionPolicy(**protection_policy),
+        "host_access_policies": [purefusion.HostAccessPolicy(**host_access_policies)],
```
Comment (andreepyro) [inline tests/functional/test_fusion_volume.py:129]: No need to add actual objects. In fact, if you want to use correct values, you should use `StorageClassRef`, `PlacementGroupRef` etc. [Here](https://github.com/PureStorage-OpenConnect/fusion-python-sdk/blob/main/fusion/models/volume.py) you can check actual parameters of the Volume object.

> swagger_types = {
        'size': 'int',
        'tenant': 'TenantRef',
        'tenant_space': 'TenantSpaceRef',
        'storage_class': 'StorageClassRef',
        'protection_policy': 'ProtectionPolicyRef',
        'placement_group': 'PlacementGroupRef',
        'array': 'ArrayRef',
        'created_at': 'int',
        'source_volume_snapshot': 'VolumeSnapshotRef',
        'host_access_policies': 'list[HostAccessPolicyRef]',
        'serial_number': 'str',
        'target': 'Target',
        'time_remaining': 'int',
        'destroyed': 'bool'
    }

I suggest using mocked string value like `"this_is_mock"` or something similar so you don't have hundreds of lines here.

File: tests/functional/test_fusion_volume.py [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+        "tenant_space": "ts1",
+        "availability_zone": "az1",
+        "storage_service": "ss1",
+        "self_link": "self_link",
+        "id": "id_1",
+    }
+
+
+@pytest.fixture
+def protection_policy():
```
Comment (andreepyro) [inline tests/functional/test_fusion_volume.py:97]: No need to change this, but JFYI you can use this too:
```
purefusion.ProtectionPolicy(
        id="098",
        name="pp1",
        self_link="self_link_value",
        display_name="Protection Policy 1",
        objectives=[],
)
```

File: tests/functional/test_fusion_volume.py [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+        fusion_volume.main()
+    assert exception.value.args[0]["changed"] is False
+
+
+@patch("fusion.OperationsApi")
+@patch("fusion.VolumesApi")
+def test_volume_create_successfully(volumes_api, operations_api, module_args):
+    operations_api = purefusion.OperationsApi()
+    volumes_api = purefusion.VolumesApi()
+    volumes_api.get_volume = MagicMock(side_effect=purefusion.rest.ApiException)
```
Comment (andreepyro) [inline tests/functional/test_fusion_volume.py:242]: Please make sure that `get_volume` method was actually called. The same for the rest of the tests, whenever something is mocked, we can take advantage of checking how many times it was called or if it was called at all. If `get_volume` wasn't called, then this test should fail as the module didn't make sure whether the volume have existed before.

The same for `get_operation` method.

File: plugins/modules/fusion_volume.py (hunk 2 above)
Comment (dimleevik) [reply plugins/modules/fusion_volume.py]: @jankodera-ps But anyways, I am not agree with empty string, if we have in ansible task empty string in size field, then it is user's problem, if user didn't set size and it thrown an error, then it is a bug. The check should be done only for null value

File: tests/functional/test_fusion_volume.py [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+    volumes_api.create_volume = MagicMock(return_value=OperationMock(1))
+    operations_api.get_operation = MagicMock(return_value=SuccessfulOperationMock)
+    set_module_args(module_args)
+    # run module
+    with pytest.raises(AnsibleExitJson) as exception:
+        fusion_volume.main()
+    assert exception.value.args[0]["changed"] is True
+    volumes_api.create_volume.assert_called_once()
+    volumes_api.create_volume.assert_called_with(
```
Comment (andreepyro) [inline tests/functional/test_fusion_volume.py:282]: These two lines can be merged into one: `assert_called_once_with(...)`. The same for all other tests.

**Review by andreepyro — CHANGES_REQUESTED** [review]

**Review by andreepyro — DISMISSED** [review]

File: plugins/modules/fusion_volume.py (hunk 2 above)
Comment (jankodera-ps) [reply plugins/modules/fusion_volume.py]: Yes, agreed, the contentious point was when size is not present at all

**Review by jankodera-ps — DISMISSED** [review]

**Review by andreepyro — DISMISSED** [review]

**Review by jankodera-ps — APPROVED** [review]

**Pull request merged**